node/app/routes.ts

```diff
@@ -0,0 +1,276 @@
+		}
+	}
+	line_commits.shift();
+	/// ^^ blame lines are 1-indexed.
+	context.data.line_commits = line_commits.map(x => {
+		if (x) {
+			return `<a href="/${context.repoName}/commit/${x}">${ x.tostrS().substr(0, 7) }</a>`;
+		} else {
+			return `&nbsp;`
+		}
+	}).join("\n");
+	
+	res.render('blame_blob', {
+		info: await TemplateInfo.info(context, req.path),
+		context,
+		line_commits,
+		layout: 'base',
+	});
+}
+//#endregion
+
+
+export const viewCommit: express.RequestHandler = async function(req, res) {
+	const context = new CommitContext(req);
+	try {
+		await context.initialize();
+	} catch(err) {
+		if (err instanceof NotFoundError) {
+			return res.status(404).send(`Not Found: ${err}`);
+		}
+	}
+	
+	const diffList = await context.commit.getDiff();
+	const diffStats = await Promise.all(diffList.map(async x => {
+		const stats = await x.getStats();
+		return `${stats.filesChanged()} changed files with ${stats.insertions()} additions and ${stats.deletions()} deletions`;
+	}));
+	
+	/// copy/paste of
+	/// github.com/nodegit/nodegit/blob/master/examples/diff-commits.js
+	let outputDiff = "";
+	for (const [i_diff, diff] of diffList.entries()) {
+		outputDiff += `Diff #${i_diff}\n`;
+		const patches = await diff.patches();
+		for (const [i_patch, patch] of patches.entries()) {
+			const hunks = await patch.hunks();
+			for (const [i_hunk, hunk] of hunks.entries()) {
+				const lines = await hunk.lines();
+				outputDiff += `\n<strong>Patch #${i_patch}, Hunk #${i_hunk}</strong>\n`;
+				outputDiff += `diff ${patch.oldFile().path()} ${patch.newFile().path()}\n`;
+				outputDiff += hunk.header().trim() + '\n\n';
+				
+				outputDiff += `<small style="color: grey">` + lines.map(line => {
+					return String.fromCharCode(line.origin()) + line.content().trim();
+				}).join('\n') + `</small>\n`;
+			}
+		}
+	}
+	
+	res.render('view_commit', {
+		info: await TemplateInfo.info(context, req.path),
+		context,
+		diffStats,
+		outputDiff,
+		no_branch_selector: true,
+		layout: 'base',
+	});
+}
+
+
+export const historyCommits: express.RequestHandler = async function(req, res) {
+	const context = new HistoryContext(req);
+	try {
+		await context.initialize();
+	} catch(err) {
+		if (err instanceof NotFoundError) {
+			return res.status(404).send(`Not Found: ${err}`);
+		}
+	}
+	
+	const revWalk = context.repo.createRevWalk();
+	revWalk.push(context.commit.id());
+	const oids: Git.Oid[] = [];
+	while (true) {
+		try {
+			const oid = await revWalk.next();
+			oids.push(oid);
+		} catch(err) {
+			if (err.errno === Git.Error.CODE.ITEROVER) {
+				break;
+			}
+			throw err;
+		}
+	}
+	const allCommits = await Promise.all(oids.map(x => context.repo.getCommit(x)));
+	/// ^^ this is very fast.
+	/// no pagination (we haven't had the use for now.)
+	let commits: Git.Commit[] = [];
+	if (context.path === undefined) {
+		commits = allCommits;
+	} else {
+		/// CAUTION(this can be pretty long)
```

Comment:
A simple way to speed this up is by not fetching the entire diff, but comparing the parents' trees to the commits' trees (without actually looking at the file contents). Looking at `.coveragerc` history takes around 30 s for me.

Reply:
See https://github.com/jonashaag/klaus/blob/purepy-hist/klaus/repo.py#L199 for an implementation

Reply:
`.coveragerc` in which repo?

Reply:
(and 30s vs. how long with this current implem?)

Yes I think you make a good point. Will have to experiment and re-implement.

Reply:
Sorry, should've been more specific. I'm talking about `huggingface/transformers/commits/master/.coveragerc`. It takes around 30 s, vs `time git log .coveragerc` => 145 ms

Reply:
In Klaus I implemented it simply by calling `git log` using subprocess. I used the pure Python implementation linked above but it was very slow.

Reply:
Ok, will take a stab at this now (unless you want to do it? :) )

Reply:
No, go for it... I invested 20 min into coming up with a draft but realised it's going to take me much longer than someone proficient in TypeScript

Reply:
I also don't know if what I came up with in Klaus is the best way to do it (but it works)

Reply:
Interesting SO question here: https://stackoverflow.com/questions/21178383/how-is-gits-log-filename-implemented-internally